pkg/template/processor_test.go

```diff
@@ -0,0 +1,345 @@
+		require.Len(t, objs, 2)
+
+		// project request
+		verifyResource(t, objs[0].Object.GetObjectKind(), project, commit, user)
+
+		// role binding
+		verifyResource(t, objs[1].Object.GetObjectKind(), project, user)
+	})
+
+	t.Run("template with missing required params", func(t *testing.T) {
+		// given
+		s := addToScheme(t)
+
+		values := make(map[string]string)
+
+		cl := test.NewFakeClient(t)
+		p := template.NewProcessor(cl, s, logger(t))
+
+		// when
+		objs, err := p.Process(templateContent(projectRequestObj, roleBindingObj), values)
+
+		// then
+		require.Error(t, err, "fail to process as not providing required param PROJECT_NAME")
+		assert.Nil(t, objs)
+	})
+
+}
+
+func TestProcessAndApply(t *testing.T) {
+	t.Run("project Request - ok", func(t *testing.T) {
+		//given
+		s := addToScheme(t)
+		project, commit, user := templateVars()
+
+		values := map[string]string{
+			"PROJECT_NAME": project,
+			"COMMIT":       commit,
+			"USER_NAME":    user,
+		}
+
+		cl := test.NewFakeClient(t)
+		p := template.NewProcessor(cl, s, logger(t))
+
+		//when
+		err := p.ProcessAndApply(templateContent(projectRequestObj), values)
+
+		//then
+		require.NoError(t, err)
+		verifyProjectRequest(t, cl, project)
+	})
+
+	t.Run("role binding - ok", func(t *testing.T) {
+		//given
+		s := addToScheme(t)
+		project, commit, user := templateVars()
+
+		values := map[string]string{
+			"PROJECT_NAME": project,
+			"COMMIT":       commit,
+			"USER_NAME":    user,
+		}
+
+		cl := test.NewFakeClient(t)
+		p := template.NewProcessor(cl, s, logger(t))
+
+		//when
+		err := p.ProcessAndApply(templateContent(roleBindingObj), values)
+
+		//then
+		require.NoError(t, err)
+		verifyRoleBinding(t, cl, project)
+	})
+
+	t.Run("project request role binding - ok", func(t *testing.T) {
+		//given
+		s := addToScheme(t)
+		project, commit, user := templateVars()
+
+		values := map[string]string{
+			"PROJECT_NAME": project,
+			"COMMIT":       commit,
+			"USER_NAME":    user,
+		}
+
+		cl := test.NewFakeClient(t)
+		p := template.NewProcessor(cl, s, logger(t))
+
+		//when
+		err := p.ProcessAndApply(templateContent(projectRequestObj, roleBindingObj), values)
+
+		//then
+		require.NoError(t, err)
+		verifyProjectRequest(t, cl, project)
+		verifyRoleBinding(t, cl, project)
+	})
+}
+
+func addToScheme(t *testing.T) *runtime.Scheme {
+	s := scheme.Scheme
+	err := apis.AddToScheme(s)
+	require.NoError(t, err)
+	utilruntime.Must(apitemplate.Install(s)) // see https://github.com/openshift/oc/blob/master/cmd/oc/oc.go#L77
+	return s
+}
+
+func templateContent(objs ...string) []byte {
+	tmpl := templateMetadata
+	for _, obj := range objs {
+		tmpl += obj
+	}
+
+	return []byte(tmpl + templateParams)
+}
+
+func logger(t *testing.T) logr.Logger {
+	zapLog, err := zap.NewDevelopment()
+	require.NoError(t, err)
+	return zapr.NewLogger(zapLog)
+}
+
+func templateVars() (string, string, string) {
+	return uuid.NewV4().String(), uuid.NewV4().String(), uuid.NewV4().String()
```

Comment:
It would be good to add small prefix, like `prj-<uuid>`, `sha-<uuid>`, 'user-<uuid>', this will help in debugging failed test case in future and add readability to test data.